### DISCUSSION TITLE
Connect relations on create doesn't seem to work (for me)

### DISCUSSION THREAD

**Discussion by matthijn** [prisma/prisma#6291] [Q&A]
I don't think this is a bug, that's why I didn't file it under that. I may just be doing things wrong.

I have the following (simplified) schema.

```prisma
model Tag {
  id String @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  value String
  jobs Job[]
}

model Job {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  name        String

  tags Tag[]
}
```

I ran `npx prisma generate && npx prisma db push --preview-feature` to generate the client and DB schema. 

Then I made a simple file to seed some data. Which works and I can view tags and jobs in `npx prisma studio`. Also in there, I can connect existing tags to existing jobs. 

But when I try to do the same in code, it doesn't work for me (the IDE also complains). 

Basically I'm doing this.

```ts
const tagIds = await Promise.all(
    tags.map(async (value) => db.tag.create({ data: {
        value
    } }))
  )

await db.job.create({
        data: {
          name: faker.lorem.lines(1),

          tags: {
            connect: tags
          }
        },
      })
```

The error I get is.

```
Type '{ name: string; tags: { connect: Tag[]; }; }' is not assignable to type 'Without<JobCreateInput, JobUncheckedCreateInput>'.       Types of property 'tags' are incompatible.         Type '{ connect: Tag[]; }' is not assignable to type 'undefined'.
```

I read the connect instructions from [here](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#connect). Where I see it both as an array or not an array to connect multiple (Tag in my case has a property id, so this should just work ,I also tried by mapping them to `[{ id: "tagId" }]` format. But that doesn't seem to change things either.

I can create the tags and jobs individually with code, just the connect part, doesn't work for me.

**Comment by ryands17** [comment]
@Matthijn 👋 

In this case, you have passed the `tags` variables which is incorrect. You need to pass the `createdTags` `id` here.

```ts
  const createdTags = await Promise.all(
    tags.map((value) =>
      prisma.tag.create({
        data: {
          value,
        },
      })
    )
  )

  await prisma.job.create({
    data: {
      name: Math.random().toString(),
      tags: {
        connect: createdTags.map((t) => ({ id: t.id })),
      },
    },
  })
```

Another better way to do this is in a single call via [nested creates](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#create) as follows:

```ts
await prisma.job.create({
    data: {
      name: Math.random().toString(),
      tags: {
        create: tags.map((value) => ({ value })),
      },
    },
  })
```

**Reply by matthijn** [reply]
Hi 👋, thanks for the reply. In this sample I indeed by accident seem to have send the tags variable. Passing the createdTags example is what I have and gives the same error.

The second sample you gave with create, just to try gives the following error.

```
      Types of property 'tags' are incompatible.
        Type '{ create: { value: string; }[]; }' is not assignable to type 'undefined'.
```

So similar. Also I think that does something a bit different right, since I want to connect existing tags to jobs, now create new tags for specific jobs.

**Reply by ryands17** [reply]
Could you share the `tags` variable that you're using?

> So similar. Also I think that does something a bit different right, since I want to connect existing tags to jobs, now create new tags for specific jobs.

You are creating tags in the command above which I thought are of the same API. If you're creating tags in a different API then `connect` is the way to go. Else, a nested `create` is a much better option.

**Comment by matthijn** [comment]
I removed all code which is not either from the `job` or `tags` table. 

So my schema.prisma is.

```prisma
model Tag {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  value     String
  jobs      Job[]
}

model Job {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  name        String

  tags Tag[]

}
```

My seed code is.

```ts
import db from "./index"
import faker from "faker"

const seed = async () => {
  const tags = [
   "tag1", "tag2"
  ]

  const createdTags = await Promise.all(
    tags.map(async (value) =>
      db.tag.create({
        data: {
          value,
        },
      })
    )
  )

  for (let j = 0; j < 5; j++) {
    // const chosenTags = sampleSize(createdTags, 3)

    await db.job.create({
      data: {
        name: faker.lorem.lines(1),

        // tags: {
        //   connect: createdTags
        // }
      },
    })
  }
}
```

After running migration en db generation `DATABASE_URL=file:./db.sqlite npx prisma migrate && DATABASE_URL=file:./db.sqlite npx prisma db push --preview-feature` and running the seed script, the above now runs. (See that I have commented out tags). But as soon as I uncomment the tags connection section, it fails. Though, the IDE is no longer complaining.

```
✕ Couldn't run imported function, are you sure it's a function?
✕ Could not seed database:
✕ Error:
Invalid `prisma.job.create()` invocation:

{
  data: {
    name: 'Ratione quam est est aut est et.',
    tags: {
    ~~~~
      connect: [
        {
          id: 'ckmwexnm90004uu43khmnsmsi',
          createdAt: new Date('2021-03-30T19:28:15.825Z'),
          updatedAt: new Date('2021-03-30T19:28:15.825Z'),
          value: 'tag1'
        },
        {
          id: 'ckmwexnm80000uu43bdrjskv6',
          createdAt: new Date('2021-03-30T19:28:15.824Z'),
          updatedAt: new Date('2021-03-30T19:28:15.824Z'),
          value: 'tag2'
        }
      ]
    }
  }
}

Unknown arg `tags` in data.tags for type JobUncheckedCreateInput. Did you mean `name`? Available args:
type JobUncheckedCreateInput {
  id?: String
  createdAt?: DateTime
  updatedAt?: DateTime
  name: String
}
```

When looking at the sqlite file, I do see the `Job`, `Tag` and `_JobToTag` table being generated.

**Reply by tarazena** [reply]
Can you check the generate Prisma generated file if it have `JobCreateInput` ?

**Reply by matthijn** [reply]
Yes, that seems to look like this.

```
export type JobCreateInput = {
    id?: string
    createdAt?: Date | string
    updatedAt?: Date | string
    name: string
    tags?: TagCreateNestedManyWithoutJobsInput
  }
```

**Reply by ryands17** [reply]
@Matthijn 👋 

As I had mentioned earlier, you cannot directly pass the tags created to `connect`. You just need to pass the `id`. Have a look a this to have a better idea.

[test-prisma.zip](https://github.com/prisma/prisma/files/6235237/test-prisma.zip)

**Reply by matthijn** [reply]
Ah, I see. I passed more information into connect besides the `id` field. I thought, it would work because `Tag` also has an `id` field. 

So, 

```
model Tag {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  value     String
  jobs      Job[]
}
``` 

[quacks like](https://en.wikipedia.org/wiki/Duck_test) `{ id: String }` too.

Perhaps a feature request then, if `{ id: xxxx }` is the only way to use that `connect` if passing in a tag. You can have something like `pick('id', itemToConnect)` in the connect code? Then you can pass anything that has an id, even if it has more fields, and thus pass a related database instance directly. 

Maybe that's not possible, I'm not sure. But this issue has been resolved. Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
